src/check_presence.c

```diff
@@ -0,0 +1,533 @@
+  }
+
+  int rc = mdb_env_create(&env);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_env_create failed: %s (%d)",
+          mdb_strerror(rc), rc);
+    return -1;
+  }
+
+  rc = mdb_env_open(env, state_datastore, MDB_NOSUBDIR, 0664);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: opening path '%s' failed: %s (%d)",
+          state_datastore, mdb_strerror(rc), rc);
+    mdb_env_close(env);
+    env = NULL;
+    return -1;
+  }
+
+  return start_thread();
+}
+
+static int presence_shutdown(void) {
+
+  if (env) {
+
+    INFO(PLUGIN_NAME " plugin: shutting down thread.");
+    if (stop_thread() < 0)
+      return -1;
+
+    mdb_env_close(env);
+  }
+
+  return 0;
+}
+
+static int presence_write(const data_set_t *ds, const value_list_t *vl,
+                          __attribute__((unused)) user_data_t *user_data) {
+
+  /* step aside quitely if not configured */
+  if (!env) {
+    return 0;
+  }
+
+  struct timeval tv_now = {0};
+
+  int rc = gettimeofday(&tv_now, /* struct timezone = */ NULL);
+  if (rc != 0) {
+    ERROR(PLUGIN_NAME " plugin: gettimeofday failed: %s", STRERRNO);
+    return -1;
+  }
+
+  /*
+   * Fast path - have we seen this host before?
+   *
+   * WHile this step is not strictly necessary, in the vast majority of
+   * cases we will have seen the host before, and we will have seen the
+   * host many times in the same second as each metric is written.
+   *
+   * The LMDB key value store offers very cheap lock free reads, allowing
+   * multiple threads to handle writes without any mutexes.
+   *
+   * This first step discards unnecessary writes as quickly as possible.
+   */
+
+  MDB_val key = {0}, data = {0};
+
+  key.mv_size = sizeof(vl->host);
+  key.mv_data = (void *)vl->host;
+
+  MDB_txn *txn = NULL;
+
+  rc = mdb_txn_begin(env, NULL, MDB_RDONLY, &txn);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_txn_begin returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    return -1;
+  }
+
+  MDB_dbi dbi = {0};
+
+  rc = mdb_dbi_open(txn, NULL, 0, &dbi);
+  if (MDB_NOTFOUND == rc) {
+
+    /* brand new database, skip the get */
+
+  } else if (rc) {
+
+    /* error happened */
+
+    ERROR(PLUGIN_NAME " plugin: mdb_dbi_open returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    mdb_txn_abort(txn);
+    return -1;
+
+  } else {
+
+    /* existing database, do the get */
+
+    rc = mdb_get(txn, dbi, &key, &data);
+  }
+
+  int add = 0;
+
+  if (MDB_NOTFOUND == rc) {
+
+    /* new host found, send a notification */
+
+    add = 1;
+
+  } else if (rc) {
+
+    /* error happened */
+
+    ERROR(PLUGIN_NAME " plugin: mdb_get returned: %s (%d)", mdb_strerror(rc),
+          rc);
+    mdb_dbi_close(env, dbi);
+    mdb_txn_abort(txn);
+    return -1;
+
+  } else {
+
+    /* have we seen host in the same second as we did our last put? if so
+     * it is good enough, exit cheaply
+     */
+
+    struct timeval *tv_then = data.mv_data;
+
+    if (data.mv_size == sizeof(struct timespec) && ((tv_then = data.mv_data)) &&
```

Comment:
Something strange I see here in `&& ((tv_then = data.mv_data))`.

Can you please carefully and critically review your code yourself?

Reply:
It'll be another week or two before I can work on this code again.

In the mean time can you fully describe what "something strange" means on that line?

In C a single "=" is an assignment operator, and in this case we don't dare try to assign the variable until we've checked the length. This is a defensive technique - don't assign possible invalid values to variables, not even temporarily.

The double brackets are a signal to the compiler that the assignment is intended.

Reply:
Fixed.

For the future, don't assume the person isn't word blind to the thing you see. @dago 's review highlighted the problem.